[HUDI-7279] make sampling rate configurable for BOUNDED_IN_MEMORY executor type

waitingF opened the pull request:
### Change Logs

When writing data with BoundedInMemoryQueue, the size of record is sampled periodically and BoundedInMemoryQueue is adjusted dynamically during runtime. Prior to this PR, the sampling rate is fixed and not configurable.

As I tested, about 9% CPU is used to estimate the object size to calculate the queue size. 

Currently the sampling rate is set to 64. If we can set the rate to 128, then we can save 4.5% CPU theoretically.
Note: Increasing sampling rate has diminishing returns beyond a certain threshold depending on record layout/schema.

![image](https://github.com/apache/hudi/assets/19326824/ff75623e-0d68-4cbf-adbc-66923683e552)


### Impact

improve performance by setting higher sampling rate

### Risk level (write none, low medium or high below)

low 

### Documentation Update

- Add a new configuration named `hoodie.write.buffer.record.sampling.rate`, default 64. Higher value leads to lower CPU usage for object size estimation.
- Add a new config named `hoodie.write.buffer.record.cache.limit` for better memory footprint control, default 128*1024.

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

codope commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -353,6 +353,18 @@ public class HoodieWriteConfig extends HoodieConfig {
+  public static final ConfigProperty<String> WRITE_BUFFER_RECORD_SAMPLING_RATE = ConfigProperty
+      .key("hoodie.write.buffer.record.sampling.rate")
+      .defaultValue(String.valueOf(64))
+      .markAdvanced()
```
Please also add `.sinceVersion("1.0.0")` for both the configs

waitingF replied:
sure, done

codope reviewed (commented):
Looks good, left one minor comment

codope reviewed (commented):
Thanks for addressing the comments. There are a couple of test failures in `TestHoodieSparkSqlWriter#testDeletePartitionsV2` and `#testSchemaEvolutionForTableType` due to OOM. Can you please check?
```
Error:  Tests run: 52, Failures: 0, Errors: 5, Skipped: 0, Time elapsed: 1,036.092 s <<< FAILURE! - in org.apache.hudi.TestHoodieSparkSqlWriter
Error:  testDeletePartitionsV2{boolean}[1]  Time elapsed: 110.88 s  <<< ERROR!
org.apache.spark.SparkException: 
Job aborted due to stage failure: Task 2 in stage 76.0 failed 1 times, most recent failure: Lost task 2.0 in stage 76.0 (TID 128, fv-az1501-788.rnq23jqhr0re1ds00u55qn22fh.cx.internal.cloudapp.net, executor driver): org.apache.spark.memory.SparkOutOfMemoryError: Unable to acquire 92 bytes of memory, got 0
```

waitingF commented:
> Thanks for addressing the comments. There are a couple of [test failures](https://github.com/apache/hudi/actions/runs/7447573282/job/20260155012?pr=10459) in `TestHoodieSparkSqlWriter#testDeletePartitionsV2` and `#testSchemaEvolutionForTableType` due to OOM. Can you please check?
> 
> ```
> Error:  Tests run: 52, Failures: 0, Errors: 5, Skipped: 0, Time elapsed: 1,036.092 s <<< FAILURE! - in org.apache.hudi.TestHoodieSparkSqlWriter
> Error:  testDeletePartitionsV2{boolean}[1]  Time elapsed: 110.88 s  <<< ERROR!
> org.apache.spark.SparkException: 
> Job aborted due to stage failure: Task 2 in stage 76.0 failed 1 times, most recent failure: Lost task 2.0 in stage 76.0 (TID 128, fv-az1501-788.rnq23jqhr0re1ds00u55qn22fh.cx.internal.cloudapp.net, executor driver): org.apache.spark.memory.SparkOutOfMemoryError: Unable to acquire 92 bytes of memory, got 0
> ```

Seems the test failure was introduced by other PR, fixed by merging master.

bvaradar reviewed (approved):
LGTM
